### PR TITLE
Create a gem which automatically configures Sidekiq.

### PR DESCRIPTION
**Pull request by elliotcm**
https://trello.com/c/u2Ba1O4W/719-create-a-gem-to-consolidate-sidekiq-usage-request-tracing-follow-on-work

This injects a Rails initializer which will
automatically configure Sidekiq for 12 factor apps
(i.e. apps that use environment variables to set
their redis configuration).

It includes our current optimal sidekiq configuration
including Airbrake/Errbit, statsd, JSON logging,
and passthrough of the `govuk_request_id`.

I don't really know how best to test this.

### CODE REVIEW

**Comment by tijmenb** [comment]
This looks excellent!

> I don't really know how best to test this.

Perhaps extract [the code from the Railtie](https://github.com/alphagov/govuk_sidekiq/blob/3204535efd5c887f8bab71761da7b8304051a3d3/lib/govuk_sidekiq/railtie.rb#L6-L32) into a method and run that? That way the code is at least exercised.

**Comment by elliotcm** [comment]
I'll see what I can do.

File: lib/govuk_sidekiq/api_headers.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+module GOVUKSidekiq
```
Comment (tijmenb) [inline lib/govuk_sidekiq/api_headers.rb]: I think `GovukSidekiq` would be more consistent with other gems. Most apps use `Govuk`: `GovukMessageQueueConsumer` / `GovukContentSchemaTestHelpers` / `GovukAdminTemplate` / `GovukTemplate`.

File: lib/govuk_sidekiq/railtie.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+          chain.add Sidekiq::Statsd::ServerMiddleware, env: "govuk.app.#{app_name}", prefix: "workers"
+          chain.add GOVUKSidekiq::APIHeaders::ServerMiddleware
+        end
+      end
+
+      Sidekiq.configure_client do |config|
+        config.redis = redis_config
+
+        config.client_middleware do |chain|
+          chain.add GOVUKSidekiq.APIHeaders::ClientMiddleware
```
Comment (tijmenb) [inline lib/govuk_sidekiq/railtie.rb]: Is there a reason to use the `.` here instead of `::`?

File: lib/govuk_sidekiq/api_headers.rb (hunk 1 above)
Comment (elliotcm) [inline lib/govuk_sidekiq/api_headers.rb]: Boo ok.

File: lib/govuk_sidekiq/railtie.rb (hunk 2 above)
Comment (elliotcm) [inline lib/govuk_sidekiq/railtie.rb]: Yes, I am a bad programmer.

**Comment by tijmenb** [comment]
@elliotcm extracting the initialiser code is something we'd have to do anyway to use it in Rummager (not a Rails app).

File: lib/govuk_sidekiq/railtie.rb (hunk 2 above)
Comment (elliotcm) [inline lib/govuk_sidekiq/railtie.rb]: (I've been writing too much Elixir)

**Comment by elliotcm** [comment]
@tijmenb Yes, sorry I didn't mean to suggest it wasn't worth doing, I'll do that plus a bit more.

**Comment by tijmenb** [comment]
👍 great, let's try this.

**Comment by elliotcm** [comment]
@tijmenb Beware, if you use this with an app which has a different GOVUK_APP_NAME to its current redis namespace you'll have to do some careful deploying to avoid losing sidekiq jobs in progress.

The apps this applies to are publishing API, signon, and the whole specialist publisher triple app nonsense.

**Pull request merged**